# Frame-Shift-Detector

# Introduction

This README contains the instructions for running the frameshift detector and helper scripts to aid in generating new input files

# Section 1. Running the Frameshift Detector #

## 1.1 The Starts File

An example file is included: starts_combined_dataset_above_90

The expected format for this input file is three lines for each gene:  
Gene Name  
Gene Nucletide Sequence  
A comma separated list of integers, corresponding to the the number of ribosome profiling fragment starts at each nucleotide of the gene

Section 3 will describe the process of generating the starts file 

## 1.2 Get genome-wide proportions of starts in frame 1 vs. frame 2 + frame 3

Get the p (f2 + f3 proportion) and q (f1 proportion) values from the starts file 

An example run:&nbsp;&nbsp;&nbsp;
python get_mean_and_std_f1.py starts_combined_dataset_above_90

will give the following output:  
F1 mean with statistics is 0.8763924094763768  
F1 stdev with statistics is 0.22906614440659637  
Total f1 is 0.914418446440859  
Total f2 is 0.04545015288643027  
Total is: 48066725

q here is 0.914, and p is (1-0.914) = 0.086

these values should be entered at the top of the frame_shift_detector.py file, see section 1.3. 

## 1.3  Run Frameshift Detector

at the top of frame_shift_detector.py  
update the following variables (p, q, p_f2, p_f3) based on the the output from get_mean_and_std_f1.py, see section 1.2

python frame_shift_detector.py starts_combined_dataset_above_90

## 1.4 The Output of Frameshift Detector

an output file is generated by appending "_with_found_frameshifts" to the input file name  

see starts_combined_dataset_above_90_with_found_frameshifts

the output contains the following information for every gene  
Gene: the name  
p-value: this is actually the exponent of the p-value and should be interpreted as 10^exponent  
gene-length:   
shift-start-position: the position relative to the start of the gene  
shift-length:   
direction: plus or minus  
before_gene_start: if yes, this frameshift should not count as plus or minus, as it is likely an un-annotated alternative gene start  
percent_shifted: into alternative frames  
num_reads: number of starts used in the p-value calculation 

